### PR TITLE
Bfb/unit testing

### PR DESCRIPTION
**Pull request by brendanburkhart**

### CODE REVIEW

File: joysticks.py [hunk 1]
```diff
@@ -1,29 +1,38 @@
...
+        self._pygame = pygame
+
+        self._pygame.init()
+
+        self._scan_joysticks()
+
+    def _scan_joysticks(self):
+        # List of all joysticks plugged in (at time of initialization)
+        self._joysticks = [self._pygame.joystick.Joystick(
+            x) for x in range(self._pygame.joystick.get_count())]
```
Comment (howardabrams) [inline joysticks.py]: Nice comprehension. Bit of a nit, but shouldn't we have `Joystick(x)` all together? Having `x` on its own looks a bit too important. ;-)

File: joysticks.py [hunk 2]
```diff
@@ -1,29 +1,38 @@
...
         list(map(lambda joy: joy.init(), self._joysticks))
```
Comment (howardabrams) [inline joysticks.py]: Why are we wrapping the `map` call with `list` when we aren't making a list of the joysticks? 

Mixing maps and comprehensions? I suppose we could use comprehensions for this too:

    [joy.init() for joy in self._joysticks]

File: joysticks.py (hunk 1 above)
Comment (brendanburkhart) [reply joysticks.py]: Good point, I didn't catch that.

File: joysticks.py (hunk 2 above)
Comment (brendanburkhart) [reply joysticks.py]: `map` just yields values - a little bit like lazy evaluation in Haskell. The `list()` was to force it to actually compute each value. I like using comprehensions all the way through much better, good suggestion.

**Comment by brendanburkhart** [comment]
I will add unit tests for the main driver station module later, I'm having issues with it right now. Mocking the configparser and figuring out a good way to unit test the PyDriverStation class are giving me some difficulties.

File: tests/test_network.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+class TestNetwork(unittest.TestCase):
+    """Test class for `Network`"""
+
+    def test_connected(self):
+        """Test that `connected` correctly reports connection status"""
+        networktables_mock = unittest.mock.Mock()
+        networktables_mock.isConnected.side_effect = [True, False]
+
+        network_instance = network.Network(networktables_mock, None, None)
```
Comment (howardabrams) [inline tests/test_network.py:17]: Nice work! You really went to town on these tests, and an effective use of mocks to boot!

File: tests/test_joysticks.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        expected_joy_data = {
+            'axes': [0.0, 1.0, -1.0],
+            'buttons': [0, 1, 1]
+        }
+
+        # Make sure joystick is initialized before being used
+        self.assertTrue(joystick_mock.init.called)
+
+        with self.assertRaises(ValueError):
+            stick.get_joystick(3)
```
Comment (howardabrams) [inline tests/test_joysticks.py:67]: Excellent! You remembered the negative tests!

File: tests/test_joysticks.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import joysticks
+
+
+class TestJoysticks(unittest.TestCase):
+    """Test class for `Joysticks`"""
+
+    def test_get_num_joysticks(self):
+        """Test whether `Joysticks.get_num_joysticks` reports
+         the correct number of joysticks
+        """
```
Comment (howardabrams) [inline tests/test_joysticks.py:15]: Function documentation on even the test functions ... really good!

/me wipes away a tear of joy.

**Review by howardabrams — APPROVED** [review]
When I suggested getting some tests in place, I didn't realize how many you were going to write. 

Very comprehensive. Really shaping up!

Next step is to incorporate the test run (and maybe even code coverage) into our CI system (Travis).

**Comment by calebeby** [comment]
Good job @BrendanBurkhart! Can we delete this branch now?

**Comment by brendanburkhart** [comment]
Yeah

**Pull request merged**